Adds 'How To Create a Starter Kit' tutorial to the docs (#83)
OpenZeppelin/starter-kit: docs/modules/ROOT/pages/create.adoc

== How to Create a Starter Kit
In this guide, you'll learn how to create your own starter kit. Not sure what starter kits are? Check out https://docs.openzeppelin.com/starter-kits/2.3/[getting started].

=== Why Should I Create a Starter Kit?

> It is still incredibly hard for new developers to enter the Ethereum ecosystem compare to Web2. New developers suffer from a lack of documentation, immature tools, and a huge upfront commitment. Starter kits represent a bridge between Web2 and Web3 supercharged to bootstrap developers into the Ethereum ecosystem within minutes using familiar concepts and technologies, such as React, NPM, Mocha while introducing Web3 components Infura, OpenZeppelin, Truffle, Rimble.

Reviewer: ```suggestion
It is still very hard for new developers to enter the Ethereum ecosystem compared to Web2. New developers suffer from a lack of comprehensive documentation and a huge upfront commitment. Starter kits represent a bridge between Web2 and Web3 supercharged to bootstrap developers into the Ethereum ecosystem within minutes, using familiar concepts and technologies, such as React, NPM, and Mocha, while introducing Web3 components like Infura, OpenZeppelin, Truffle, and Rimble.
```

Reviewer: It feels odd to mention immature tools and poor documentation in the documentation site of a tool

> Creating a starter kit dedicated to a particular project or tool is a great way to spotlight it for new developers in the space as well as to create hands-on experience and materials to run workshops.

Reviewer: Loved this

https://github.com/OpenZeppelin/starter-kit-gsn[GSN Starter Kit] is an example of a project focused starter kit which helped to kickstart https://gsn.openzeppelin.com/[Gas Station Network].

=== How to Create a Starter Kit

> To create an OpenZeppelin Kit you’ll need to publish a GitHub repo containing your kit project. The GitHub repo has to contain a `kit.json` configuration file and have a stable branch. The repo should have meaningful `README.md` file walking users through installation, running and building a kit. The `stable` branch will be used as a source for the kit's deployment. Use the stable branch only for kit releases. https://github.com/OpenZeppelin/starter-kit-tutorial[The tutorial kit] is a good example.

Reviewer: ```suggestion
To create an OpenZeppelin Kit you’ll need to publish a GitHub repo containing your kit project. The GitHub repo has to contain a `kit.json` configuration file and have a stable branch. The repo should have a meaningful `README.md` file walking users through installation, running and building a kit. The `stable` branch will be used as a source for the kit's deployment. Use the stable branch only for kit releases. https://github.com/OpenZeppelin/starter-kit-tutorial[The tutorial kit] is a good example.
```

.*Message*
A message displayed in the terminal immediately after installation. Keep it short and to the point.

.*Files*
An array of files and folders to be copied from the repo. An empty array would result in copying all the files.

.*Hooks*
An object containing terminal commands to execute during unpacking. Currently, only the `post-unpack` hook is supported, which is executed right after unpacking a kit.

=== How to Test a Starter Kit
Once your starter kit is ready it is always a good idea to give it a test run. It is possible to install a starter kit from any Git branch using syntax `npx openzepelin unpack 'yourGitHubHandle/yourRepoName#branch'`, for example `npx openzeppelin unpack 'OpenZeppelin/starter-kit#develop'`. Make sure your starter kit runs fine on Linux, Mac OS, and Windows. We have experienced installation and running issues depending on a platform.

=== How to Release a New Version
Once you've added desired changes to your starter kit, push them to the `stable` branch and they will be instantly available using an `unpack` command.

> === How to Install My Starter Kit

Reviewer: ```suggestion
=== How to Install a Starter Kit
```

For consistency with the subtitle "how to test..."

> == How to Create a Starter Kit

Reviewer: Make it consistent with the title on the sidebar

> === What is a Starter Kit?

Reviewer: I'd drop this section, since this is already covered previously in the docs. Do add something like this if you share this on the forum, though.

Instead, I'd add an intro paragraph (right below the main title) explaining what this doc is about: _in this guide you'll learn to create your own starter kit, read more about them here (link to getting started)_, or similar. Conversely, I would not include this paragraph in a posting on the forum, just to the documentation.

https://openzeppelin.com/starter-kits/[Starter Kit] is a preconfigured Ethereum project containing the best industry tools to bootstrap a dApp development process.

=== Why Should I Create a Starter Kit?

> Anyone can create a Starter Kit and it is a great way to introduce developers to a new tool or a project, like https://github.com/OpenZeppelin/starter-kit-gsn[GSN Starter Kit].

Reviewer: I'd expand a bit more here, especially on the ethos behind kits, in that the idea is to provide devs with a box with everything ready to hit the road running when building a project.

=== How to Create a Starter Kit

> To create an OpenZeppelin Kit you’ll need to publish a GitHub repo containing your kit project. The GitHub repo has to contain the kit.json configuration file and have a stable branch. Repo should have meaningful README.md file walking users through installation, running and building a kit. Repo should have a stable branch that is used as a source for a kit deployment. Use a stable branch only for kit releases. https://github.com/OpenZeppelin/starter-kit-tutorial[The tutorial kit] is a good example.

Reviewer: ```suggestion
To create an OpenZeppelin Kit you’ll need to publish a GitHub repo containing your kit project. The GitHub repo has to contain a `kit.json` configuration file and have a stable branch. The repo should have meaningful `README.md` file walking users through installation, running and building a kit. The `stable` branch will be used as a source for the kit's deployment. Use the stable branch only for kit releases. https://github.com/OpenZeppelin/starter-kit-tutorial[The tutorial kit] is a good example.
```

==== Checklist

- Create a GitHub repo
- Create a stable branch on the repo
- Add a valid configuration file to the root of the repo
- Add desired tools and packages to the repo
- Add meaningful README.md file to the repo
- Push a release version to a `stable` branch

==== Configuration file

> Every Starter kit includes a configuration file – `kit.json`. This file has the following structure:

Reviewer: ```suggestion
Every Starter kit must include a `kit.json` configuration file. This file has the following structure:
```

[source, json]
----
    {
      "manifestVersion": "0.1.0",
      "message": "More at https://github.com/OpenZeppelin/starter-kit",
      "files": [
        ".gitignore",
        "LICENSE",
        "client",
        "contracts",
        "migrations",
        "package.json",
        "solhint.json",
        "test",
        "truffle-config.js"
      ],
      "hooks": {
        "post-unpack": "npm install && cd client && npm install"
      }
    }
----

.*manifestVersion*

> A version of a manifest that ensures proper handling of a kit. Always use the version provided in the documentation.

Reviewer: I'd also clarify that they should **not** bump this version if they want to publish a new version of the kit, or clarify that it's the version of this configuration file and not of the project.

.*Message*

> A message displayed in the terminal immediately after installation. Keep it short and to the point.

Reviewer: > Keep it short and to the point.

Love that you are including suggestions on making a good project, such as this one or the one on the README, and not just on building the kit.

.*Files*
An array of files and folders to be copied from the repo. An empty array would result in copying all the files.

.*Hooks*

> An object containing terminal commands to execute during unpacking. Currently, only post-unpack command supported which is executed right after unpacking a kit.

Reviewer: ```suggestion
An object containing terminal commands to execute during unpacking. Currently, only the `post-unpack` hook is supported, which is executed right after unpacking a kit.
```
